### PR TITLE
[Bots] Add Caster Range Command, and IsValidSpellRange Checks

### PR DESCRIPTION
**Pull request by Aeadoin**
Casters will try to be at the edge of the "Caster Range" to prevent them from rushing into melee range.

Also added IsValidSpellRange checks for Bots where it makes sense (Nukes, DoTs, Snare, Root, Heals)

### CODE REVIEW

File: zone/bot.cpp [hunk 1]
```diff
@@ -9830,4 +9812,30 @@ void Bot::SendSpellAnim(uint16 target_id, uint16 spell_id)
...
+	float caster_distance_min = 0.0f;
+	float caster_distance = 0.0f;
+	{
+		if (!GetBotCasterRange()) {
+			if (GetLevel() >= GetStopMeleeLevel() && GetClass() >= WARRIOR && GetClass() <= BERSERKER) {
+				caster_distance_max = MAX_CASTER_DISTANCE[GetClass() - 1];
+			}
+		}
+		else {
+			caster_distance_max = GetBotCasterRange() * GetBotCasterRange();
```
Comment (Akkadius) [inline zone/bot.cpp:9826]: The else can be removed and go before the if check

```cpp
caster_distance_max = GetBotCasterRange() * GetBotCasterRange();
if (!GetBotCasterRange()) {
	if (GetLevel() >= GetStopMeleeLevel() && GetClass() >= WARRIOR && GetClass() <= BERSERKER) {
		caster_distance_max = MAX_CASTER_DISTANCE[GetClass() - 1];
	}
}
```

File: zone/bot_database.cpp [hunk 2]
```diff
@@ -3148,6 +3151,27 @@ std::string BotDatabase::GetBotNameByID(const uint32 bot_id)
...
+			"UPDATE `bot_data`"
+			" SET `caster_range` = '%u'"
+			" WHERE `owner_id` = '%u'"
+			" AND `bot_id` = '%u'",
+			bot_caster_range_value,
+			owner_id,
+			bot_id
+	);
+	auto results = database.QueryDatabase(query);
+	if (!results.Success())
```
Comment (Akkadius) [inline zone/bot_database.cpp:3169]: Close this off please

File: zone/botspellsai.cpp [hunk 3]
```diff
@@ -3456,3 +3483,17 @@ bool Bot::HasBotSpellEntry(uint16 spellid) {
...
 	return false;
 }
+
+bool Bot::IsValidSpellRange(uint16 spell_id, Mob const* tar) {
+	if (!IsValidSpell(spell_id)) {
+		return false;
+	}
+
+	if (tar) {
+		int spellrange = .8 * (GetActSpellRange(spell_id, spells[spell_id].range) * GetActSpellRange(spell_id, spells[spell_id].range));
```
Comment (Akkadius) [inline zone/botspellsai.cpp:3493]: .8 ?

File: zone/bot.cpp (hunk 1 above)
Comment (Aeadoin) [reply zone/bot.cpp]: Cleaned up.

File: zone/bot_database.cpp (hunk 2 above)
Comment (Aeadoin) [reply zone/bot_database.cpp]: With brackets you mean? if so done.

File: zone/botspellsai.cpp (hunk 3 above)
Comment (Aeadoin) [reply zone/botspellsai.cpp:3493]: Per Nite.
image

I'll do some testing, and see why this is needed, or least assign it to a local variable to give it more verbosity..

File: zone/bot_database.cpp [hunk 4]
```diff
@@ -3148,6 +3151,30 @@ std::string BotDatabase::GetBotNameByID(const uint32 bot_id)
 	return nullptr;
 }
 
+bool BotDatabase::SaveBotCasterRange(const uint32 owner_id, const uint32 bot_id, const uint32 bot_caster_range_value)
+{
+	if (!owner_id || !bot_id) {
+		return false;
+		}
```
Comment (Akkadius) [inline zone/bot_database.cpp:3158]: Formatting

File: zone/bot_database.cpp [hunk 5]
```diff
@@ -3148,6 +3151,30 @@ std::string BotDatabase::GetBotNameByID(const uint32 bot_id)
...
+			" AND `bot_id` = '{}'",
+			bot_caster_range_value,
+			owner_id,
+			bot_id
+	);
+	auto results = database.QueryDatabase(query);
+
+	if (!results.Success()) {
+		return false;
+		}
```
Comment (Akkadius) [inline zone/bot_database.cpp:3173]: Formatting

File: zone/botspellsai.cpp (hunk 3 above)
Comment (Aeadoin) [reply zone/botspellsai.cpp:3493]: Tested myself, and this wasn't required. so removed.

File: zone/bot_database.cpp (hunk 4 above)
Comment (Aeadoin) [reply zone/bot_database.cpp]: Fixed.

File: zone/bot_database.cpp (hunk 5 above)
Comment (Aeadoin) [reply zone/bot_database.cpp]: Fixed.

**Review by Akkadius — APPROVED** [review]

**Pull request merged**